File: ruby/problem002.rb
require './euler_solution.rb'

class Problem002
  include EulerSolution

  def run
    # ~0ms
    fast_fibs.take_while {|i| i < 4_000_000}.select {|i| i % 2 == 0}.sum

    # ~1500ms
    # fibs.take_while {|i| i < 4_000_000}.select {|i| i % 2 == 0}.sum
  end

  def fast_fibs
    Enumerator.new do |y|
      n1 = 0
      n2 = 1

      loop do
        y << n1
        n1, n2 = n2, n1 + n2
      end
    end
  end

  def fibs
    1.step(by: 1).lazy.map{|i| fib(i)}
  end

  def fib(n)
    return n if n == 0 || n == 1
    fib(n - 1) + fib(n - 2)
  end
end

execute(Problem002.new)

File: kotlin/Problem002.kt
fun main(args: Array<String>) {
    EulerSolution.execute(Problem002())
}

class Problem002 : EulerSolution {
    override fun run() : Long {
        // ~10ms
        return fastFibs().takeWhile{it < 4000000}.filter{it % 2 == 0}.sum().toLong()

        // ~80ms
        // return fibs().takeWhile{it < 4000000}.filter{it % 2 == 0}.sum().toLong()
    }

    fun fastFibs() : Sequence<Int> {
        // pairs mapped to single ints
        return generateSequence(Pair(0, 1), { Pair(it.second, it.first + it.second) }).map { it.first }
    }

    fun fibs() : Sequence<Int> {
        // sequence starting at 0, incrementing by 1, mapped to fibonacci number
        return generateSequence(0) {it + 1}.map(this::fib)
    }

    fun fib(n : Int) : Int {
        if (n == 0 || n == 1) {
            return n
        }
        return fib(n - 1) + fib(n - 2)
    }
}